File: Kick_Start/2020/Kick_Start_2020-H-4-2.py
# Friends
# https://codingcompetitions.withgoogle.com/kickstart/round/000000000019ff49/000000000043aee7
# Use Floyd–Warshall algorithm to find all pair of shortest paths between letters
# O(L^2*N + 26^3 + L^2*Q) = O(L^2*(N+Q))

INF = pow(10, 5)
T = int(input())
for t in range(T):
    [N, Q] = [int(n) for n in input().split()]
    names = input().split()

    dist = [[INF] * 26 for i in range(26)]
    for i in range(26):
        dist[i][i] = 0
    # Letters have edge only if they appear in same name
    for i in range(N):
        names[i] = list(set(names[i]))
        L = len(names[i])
        for j in range(L):
            for k in range(j + 1, L):
                dist[ord(names[i][j]) - ord('A')][ord(names[i][k]) - ord('A')] = 1
                dist[ord(names[i][k]) - ord('A')][ord(names[i][j]) - ord('A')] = 1
    for k in range(26):
        for i in range(26):
            for j in range(26):
                if dist[i][k] + dist[k][j] < dist[i][j]:
                    dist[i][j] = dist[i][k] + dist[k][j]

    dist_people = {}

    ans = []
    for i in range(Q):
        [x, y] = [int(n) for n in input().split()]
        if (x - 1, y - 1) not in dist_people:
            temp = INF
            for letter1 in names[x - 1]:
                for letter2 in names[y - 1]:
                    temp = min(temp, dist[ord(letter1) - ord('A')][ord(letter2) - ord('A')])
            # temp: # of edge in letter's graph, means there are temp + 1 letters are used to connect, means there are temp + 2 people in chain
            dist_people[(x - 1, y - 1)] = temp + 2 if temp != INF else -1
            dist_people[(y - 1, x - 1)] = dist_people[(x - 1, y - 1)]
        ans.append(dist_people[(x - 1, y - 1)])

    print(f"Case #{t + 1}: {' '.join([str(n) for n in ans])}")

File: Google-Kick-Start/2020/Kick_Start_2020-H-4-2.cpp
// Friends
// https://codingcompetitions.withgoogle.com/kickstart/round/000000000019ff49/000000000043aee7
// Use Floyd–Warshall algorithm to find all pair of shortest paths between letters
// O(L^2*N + 26^3 + L^2*Q) = O(L^2*(N+Q))

#include <bits/stdc++.h>

typedef long long ll;

using namespace std;

int INF = pow(10, 5);
int main() {
    int T;
    cin >> T;
    for (int t = 0; t < T; t++) {
        int N, Q;
        cin >> N >> Q;
        vector<string> names(N);
        for (int i = 0; i < N; i++)
            cin >> names[i];

        vector< vector<int> > dist(26, vector<int>(26, INF));
        for (int i = 0; i < 26; i++)
            dist[i][i] = 0;
        // Letters have edge only if they appear in same name
        for (int i = 0; i < N; i++) {
            for (char c1: names[i]) {
                for (char c2: names[i]) {
                    if (c1 != c2) {
                        dist[c1 - 'A'][c2 - 'A'] = 1;
                        dist[c2 - 'A'][c1 - 'A'] = 1;
                    }
                }
            }
        }
        for (int k = 0; k < 26; k++) {
            for (int i = 0; i < 26; i++) {
                for (int j = 0; j < 26; j++) {
                    if (dist[i][k] + dist[k][j] < dist[i][j])
                        dist[i][j] = dist[i][k] + dist[k][j];
                }
            }
        }
        
        printf("Case #%d:", t + 1);
        for (int i = 0; i < Q; i++) {
            int x, y;
            cin >> x >> y;
            int ans = INF;
            for (char c1: names[x - 1]) {
                for (char c2: names[y - 1]) {
                    ans = min(ans, dist[c1 - 'A'][c2 - 'A']);
                }
            }
            if (ans != INF)
            // ans: # of edge in letter's graph, means there are ans + 1 letters are used to connect, means there are ans + 2 people in chain
                printf(" %d", ans + 2);
            else
                printf(" -1");
        }
        printf("\n");
    }
}
